common/flatpak-run-wayland.c

```diff
@@ -94,7 +135,13 @@ flatpak_run_add_wayland_security_context_args (FlatpakBwrap *bwrap,
 
   *available_out = TRUE;
 
-  display = wl_display_connect (NULL);
+  /* In order to support WAYLAND_SOCKET, the implementation here:
+   * 1. Avoid using wl_display_connect, since it will unconditonally call
+   *    getenv and unsetenv on WAYLAND_SOCKET.
+   *    The fd wihtin WAYLAND_SOCKET is supposed to be only used by application.
+   * 2. We still set up security context for the other processes in sandbox
+   *    that do not use WAYLAND_SOCKET. */
+  display = connect_to_wayland_display ();
```

Comment:
Perhaps more like this:

```suggestion
  /* We don't use wl_display_connect() here, for two reasons:
   * 1. It would unsetenv ("WAYLAND_SOCKET"), which is not thread-safe.
   * 2. If the compositor has set WAYLAND_SOCKET to a special, higher-privileged
   *    socket, the application should be able to get those privileges for its first
   *    connection; but that fd can only be used once, so having flatpak itself
   *    do that first connection would defeat that mechanism.
   *
   * We still set up a security context for the second and subsequent connections
   * to Wayland from within the sandbox.
   */
  display = connect_to_wayland_display ();
```